enderio-base/src/main/java/com/enderio/base/common/item/darksteel/upgrades/explosive/ExplosiveUpgradeHandler.java

```diff
@@ -0,0 +1,263 @@
+        var fromCorners = getCorners(refBounds);
+        var toCorners = getCorners(miningBounds);
+        for (int i=0;i<fromCorners.size();i++) {
+            Vector3d from = fromCorners.get(i);
+            Vector3d to = toCorners.get(i);
+            addVertices(poseStack.last(), vertexConsumer, origin, from, to, color);
+        }
+    }
+
+    private static List<Vector3d> getCorners(AABB aabb) {
+        List<Vector3d> res = new ArrayList<>(8);
+        res.add(new Vector3d(aabb.minX, aabb.minY, aabb.minZ));
+        res.add(new Vector3d(aabb.minX, aabb.maxY, aabb.minZ));
+        res.add(new Vector3d(aabb.maxX, aabb.minY, aabb.minZ));
+        res.add(new Vector3d(aabb.maxX, aabb.maxY, aabb.minZ));
+        res.add(new Vector3d(aabb.minX, aabb.minY, aabb.maxZ));
+        res.add(new Vector3d(aabb.minX, aabb.maxY, aabb.maxZ));
+        res.add(new Vector3d(aabb.maxX, aabb.minY, aabb.maxZ));
+        res.add(new Vector3d(aabb.maxX, aabb.maxY, aabb.maxZ));
+        return res;
+    }
+
+    private static void renderShape(PoseStack pPoseStack, VertexConsumer pConsumer, VoxelShape pShape, Vector3d origin, Vector4f color) {
+        PoseStack.Pose pose = pPoseStack.last();
+        pShape.forAllEdges(
+            (pMinX, pMinY, pMinZ, pMaxX, pMaxY, pMaxZ) -> addVertices(pose, pConsumer, origin.x, origin.y, origin.z, pMinX, pMinY, pMinZ, pMaxX, pMaxY, pMaxZ,
+                color));
+    }
+
+    private static void addVertices(PoseStack.Pose pose, VertexConsumer pConsumer, Vector3d origin, Vector3d from, Vector3d to, Vector4f color) {
+        addVertices(pose,pConsumer, origin.x, origin.y, origin.z, from.x, from.y, from.z, to.x, to.y, to.z, color);
+    }
+
+    private static void addVertices(PoseStack.Pose pose, VertexConsumer pConsumer, double originX, double originY, double originZ,
+        double fromX, double fromY, double fromZ, double toX, double toY, double toZ, Vector4f color) {
+
+        //create normal
+        float normalX = (float)(toX - fromX);
+        float normalY = (float)(toY - fromY);
+        float normalZ = (float)(toZ - fromZ);
+        float length = Mth.sqrt(normalX * normalX + normalY * normalY + normalZ * normalZ);
+        //normalize
+        normalX = normalX / length;
+        normalY = normalY / length;
+        normalZ = normalZ / length;
+
+        pConsumer
+            .vertex(pose.pose(), (float) (fromX + originX), (float) (fromY + originY), (float) (fromZ + originZ))
+            .color(color.x(), color.y(), color.z(), color.w())
+            .normal(pose.normal(), normalX, normalY, normalZ)
+            .endVertex();
+        pConsumer
+            .vertex(pose.pose(), (float) (toX + originX), (float) (toY + originY), (float) (toZ + originZ))
+            .color(color.x(), color.y(), color.z(), color.w())
+            .normal(pose.normal(), normalX, normalY, normalZ)
+            .endVertex();
+    }
+
+    // endregion
```

Comment:
This might be fine to stay, but it could also be moved to an util class, since it seems like those are more general rendering methods.

Reply:
Yeah, I considered this but only a couple of the methods are useful, and only for drawing lines. I would prefer to move it out id we actually have another need but don't really mind